airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJdbcSource.java

```diff
@@ -104,101 +118,249 @@ public AirbyteConnectionStatus check(JsonNode config) {
+
+      final Stream<AirbyteMessage> internalMessageStream;
+      if (cursorOptional.isPresent()) {
+        internalMessageStream = getIncrementalStream(connection, airbyteStream, selectedDatabaseFields, table, cursorOptional.get(), now);
+      } else {
+        // if no cursor is present then this is the first read for is the same as doing a full refresh read.
+        internalMessageStream = getFullRefreshStream(connection, streamName, selectedDatabaseFields, table, now);
+      }
 
-  @Override
-  public Stream<AirbyteMessage> read(JsonNode config, ConfiguredAirbyteCatalog catalog, JsonNode state) throws Exception {
-    final Instant now = Instant.now();
+      final JsonSchemaPrimitive cursorType = IncrementalUtils.getCursorType(airbyteStream, cursorField);
+      final StateDecoratingIterator stateDecoratingIterator = new StateDecoratingIterator(
+          internalMessageStream,
+          stateManager,
+          streamName,
+          cursorField,
+          cursorOptional.orElse(null),
+          cursorType);
+
+      stream = MoreStreams.toStream(stateDecoratingIterator);
+    } else if (airbyteStream.getSyncMode() == SyncMode.FULL_REFRESH || airbyteStream.getSyncMode() == null) {
+      stream = getFullRefreshStream(connection, streamName, selectedDatabaseFields, table, now);
+    } else {
+      throw new IllegalArgumentException(String.format("%s does not support sync mode: %s.", airbyteStream.getSyncMode(), AbstractJdbcSource.class));
+    }
 
-    final Database database = createDatabase(config);
+    return stream;
+  }
 
-    final Map<String, TableInfo> tableNameToTable = getTables(database).stream()
-        .collect(Collectors.toMap(TableInfo::getName, Function.identity()));
+  private static Stream<AirbyteMessage> getIncrementalStream(Connection connection,
+                                                             ConfiguredAirbyteStream airbyteStream,
+                                                             List<String> selectedDatabaseFields,
+                                                             TableInfoInternal table,
+                                                             String cursor,
+                                                             Instant now)
+      throws SQLException {
+    final String streamName = airbyteStream.getStream().getName();
+    final String cursorField = IncrementalUtils.getCursorField(airbyteStream);
+    final JDBCType cursorJdbcType = table.getFields().stream()
+        .filter(info -> info.getColumnName().equals(cursorField))
+        .map(ColumnInfo::getColumnType)
+        .findFirst()
+        .orElseThrow();
+
+    Preconditions.checkState(table.getFields().stream().anyMatch(f -> f.getColumnName().equals(cursorField)),
+        String.format("Could not find cursor field %s in table %s", cursorField, table.getName()));
+
+    final ResultSet resultSet = queryTableIncremental(
```

Comment:
Do you have the same guarantee around laziness here that we had with JOOQ? If not, we're in trouble

Reply:
discussed offline. it looks like many jdbc drivers will act lazily here, but there is _not_ a guarantee. i have added lazy functionality in the [db class](https://github.com/airbytehq/airbyte/pull/1305/files) i was working on to get this guarantee.

Reply:
we should now have the guarantees we need.